ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -877,6 +963,52 @@ MessageFormatFlags getOperationFlag() {
           : MessageFormatFlags.All;
     }
 
+    @Override
+    boolean mayBeProcessDecryptCallbackAndCompleteOperation() {
+      if (onDecryptMode() && decryptCallbackResultAvailable) {
+        // incase of Metadata blob, only user-metadata needs decryption if the blob is encrypted
+        if (blobType == BlobType.MetadataBlob) {
+          if (decryptCallbackException == null) {
+            logger.trace("Processing stored decryption callback result for Metadata blob {}", blobId);
+            blobInfo = new BlobInfo(blobInfoFromServer.getBlobProperties(),
+                decryptCallbackResult.getDecryptedUserMetadata().array());
+          } else {
+            logger.error("Decryption job callback invoked with exception {} for Metadata blob {} ",
+                decryptCallbackException.getCause(), blobId);
+            setOperationException(
+                new RouterException("Exception thrown on decrypting content for Metadata blob " + blobId,
+                    RouterErrorCode.UnexpectedInternalError));
+          }
+          logger.trace("BlobContent available to process for Metadata blob {}", blobId);
+        } else {
+          // Incase of simple blobs, user-metadata may or may not be passed into decryption job based on GetOptions flag.
+          // Only in-case of GetBlobInfo and GetBlobAll, user-metadata is required to be decrypted
+          if (decryptCallbackException == null) {
+            logger.trace("Processing stored decryption callback result for simple blob {}", blobId);
+            // setting right BlobInfo if applicable
+            if (blobInfoFromServer != null) {
```

Comment:
I don't know if the blobInfoFromServer variable is needed. You could just check if the decryptCallbackResult is non-null and overwrite blobInfo in that case.

Reply:
this was added during the progress of this patch, wherein I was accepting responses from other servers until decryption is complete. But is not the case anymore. Will remove it.